File: src/main/java/myblog/blog/category/repository/CategoryRepository.java
package myblog.blog.category.repository;

import myblog.blog.category.domain.Category;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;

import java.util.List;

public interface CategoryRepository extends JpaRepository<Category, Long> {

    /*
        - 카테고리 이름으로 카테고리 찾기
    */
    Category findByTitle(String title);

    /*
        - 티어별 카테고리들 가져오기
    */
    List<Category> findAllByTierIs(int tier);

    /*
        - ID == 0 인 더미카테고리를 제외한 모든 카테고리 불러오기
    */
    @Query("select c " +
            "from Category c " +
            "where c.id >0 ")
    List<Category> findAllWithoutDummy();

}

File: service/blog/blog-core/src/main/kotlin/kr/co/jiniaslog/blog/outbound/persistence/CategoryRepository.kt
package kr.co.jiniaslog.blog.outbound.persistence

import kr.co.jiniaslog.blog.domain.category.Category
import kr.co.jiniaslog.blog.domain.category.CategoryId
import kr.co.jiniaslog.shared.core.domain.Repository
import org.springframework.data.jpa.repository.JpaRepository

interface CategoryRepository : Repository<Category, CategoryId> {
    fun findAll(): List<Category>

    fun deleteAll(list: List<Category>)

    fun saveAll(list: List<Category>): List<Category>
}

interface CategoryJpaRepository : JpaRepository<Category, CategoryId>

@org.springframework.stereotype.Repository
class CategoryRepositoryAdapter(
    private val categoryJpaRepository: CategoryJpaRepository
) : CategoryRepository {
    override fun save(entity: Category): Category {
        return categoryJpaRepository.save(entity)
    }

    override fun findAll(): List<Category> {
        return categoryJpaRepository.findAll()
    }

    override fun deleteAll(list: List<Category>) {
        return categoryJpaRepository.deleteAll(list)
    }

    override fun saveAll(list: List<Category>): List<Category> {
        return categoryJpaRepository.saveAll(list)
    }

    override fun findById(id: CategoryId): Category? {
        return categoryJpaRepository.findById(id).orElse(null)
    }

    override fun deleteById(id: CategoryId) {
        categoryJpaRepository.deleteById(id)
    }
}
